test/SendRecv_OutOfPlace.cpp

```diff
@@ -0,0 +1,90 @@
+/*************************************************************************
+ * Copyright (c) 2022 Advanced Micro Devices, Inc. All rights reserved.
+ *
+ * See LICENSE.txt for license information
+ ************************************************************************/
+#include "TestBed.hpp"
+
+namespace RcclUnitTesting
+{
+  TEST(SendRecv, OutOfPlace)
+  {
+    TestBed testBed;
+
+    // Configuration
+    std::vector<ncclFunc_t>     const  funcType        = {ncclCollSend, ncclCollRecv}; // akollias
+    std::vector<ncclDataType_t> const& dataTypes       = {ncclInt32};
+    std::vector<ncclRedOp_t>    const& redOps          = {ncclSum}; //Not important for send receive tests
+    std::vector<int>            const  numElements     = {1024}; // akollias, one test for now
+    bool                        const  inPlace         = false;
+    bool                        const  useManagedMem   = false;
+    int                         const  numCollPerGroup = numElements.size();
+
+    // CollFuncPtr prepFunc = DefaultPrepData_SendRecv; // akollias
+    bool isCorrect = true;
+    // for (int totalRanks = testBed.ev.minGpus; totalRanks <= testBed.ev.maxGpus && isCorrect; ++totalRanks) // big iterator for all ranks
+    int totalRanks = testBed.ev.maxGpus; // akollias this to change on maxGpus
+      int const numProcesses = 1;
+      int const isMultiProcess = 0;
+      testBed.InitComms(TestBed::GetDeviceIdsList(numProcesses, totalRanks), numCollPerGroup);
+
+      // for (int redOpIdx = 0; redOpIdx < redOps.size() && isCorrect; ++redOpIdx)
+      for (int dataIdx = 0; dataIdx < dataTypes.size() && isCorrect; ++dataIdx)
+      {
+        if (testBed.ev.showNames) // Show test names
+          INFO("%s process %2d-ranks SendRec v%d out Of place (%s)\n",
+               isMultiProcess ? "Multi " : "Single",
+               numCollPerGroup,
+               totalRanks, ncclDataTypeNames[dataTypes[dataIdx]]); // akollias no Red_ops
+
+        // Run all element sizes in parallel as single group // akollias for now this will only run for one test
+        for (int root = 0; root < totalRanks; ++root)
+        {
+          testBed.SetCollectiveArgs(funcType[0],
+                                    dataTypes[dataIdx],
+                                    redOps[0],
+                                    0, // 0?
+                                    numElements[0],
+                                    numElements[0],
+                                    0,
+                                    root);
+          testBed.AllocateMem(inPlace, useManagedMem, -1, root);
+          testBed.PrepareData(-1, root);
+          for (int currentRank = 0; currentRank < totalRanks; ++currentRank)
+          { // so root needs the recv number, and recv gpu needs the root rank
+
+            if (currentRank != root)
+            {
+              testBed.SetCollectiveArgs(funcType[0],

```

Comment:
This is somewhat confusing.  The initial set collectives call suggests that root will be the one that is performing the send.  However, here the current rank performs the send, while the root is the one doing the receive.  
Chances are, the first call was so that you could allocate only once, however it would be better if you just had these calls once.  

Suggestion to match other parts of the code:

[for each sendRank from 0 to ...]
{
  [for each recvRank from 0 to ...]
  {
    Set SendRank collective
    if (recvRank == 0) AllocateMem for SendRank
    if (recvRank == sendRank) continue;
    Set RecvRank collective
    AllocateMem for RecvRank
    ...
    ...
   DeallocateMem for RecvRank
   }
  DeallocateMem for SendRank
}

Reply:
The p[roblem with this is that for every case where SendRank is !=0 it will not allocate memory early enough fto run the test (i.e case SendRank = 1 ReceiveRank = 0)